### ISSUE TITLE
BUG: numpy 1.22 incorrectly infers dimensions

### ISSUE THREAD

**Issue by ketch** [numpy/numpy#21362]
### Describe the issue:

Using f2py to wrap certain routines in Clawpack, on some systems we get a different calling signature using numpy 1.22 versus earlier versions of numpy.  Some specifics are described here: https://github.com/clawpack/pyclaw/issues/681

Some of the input arrays' dimensions involve two variables, so normally f2py requires that those values be passed in.  Using 1.22, sometimes f2py sets up the wrapper so that both of those values are inferred, but their inferred values depend on each other.  So even if one passes the values in, the routine does not work correctly.

I've found that rolling back to numpy version 1.18.5 fixes this issue.  I haven't tested other versions in between. 

I wonder if this is related to https://github.com/numpy/numpy/issues/20103 or https://github.com/numpy/numpy/issues/20696.

### Reproduce the code example:

```python
After installing Clawpack, run any PyClaw example.
```

### NumPy/Python version information:

1.22.3

**Comment by charris** [comment]
ISTR another report along these lines. @HaoZeke Thoughts?

**Comment by HaoZeke** [comment]
> ISTR another report along these lines. @HaoZeke Thoughts?

Yup, there have been a couple growing pains (https://github.com/numpy/numpy/pull/20721, https://github.com/numpy/numpy/issues/20709, https://github.com/numpy/numpy/issues/20853) from our new dimension solver https://github.com/numpy/numpy/pull/19805, but the fixes have been relatively quick to implement since it makes reasoning about the code much easier than before. Will look into this soon.

**Comment by HaoZeke** [comment]
This seems to be a non-starter.
```
git clone https://github.com/clawpack/clawpack.git
cd clawpack
git submodule init
git submodule update
pip install -e .
python -c "import clawpack.pyclaw.classic.classic1 as c; print(help(c))"
```
Gives:
```
Help on module clawpack.pyclaw.classic.classic1 in clawpack.pyclaw.classic:

NAME
    clawpack.pyclaw.classic.classic1

DESCRIPTION
    This module 'classic1' is auto-generated with f2py (version:1.23.0.dev0+1087.g0eaa40db3).
    Functions:
        limiter(maxm,num_ghost,mx,wave,s,mthlim,num_eqn=shape(wave, 0),num_waves=shape(wave, 1))
        philim = philim(a,b,meth)
        q,cfl = step1(num_ghost,mx,q,aux,dx,dt,method,mthlim,use_fwave,rp1,num_eqn=shape(q, 0),num_waves=shape(mthlim, 0),num_aux=shape(aux, 0),f=,wave=,s=,amdq=,apdq=,dtdx=,rp1_extra_args=())
    .

DATA
    __f2py_numpy_version__ = '1.23.0.dev0+1087.g0eaa40db3'
    limiter = <fortran object>
    philim = <fortran philim>
    step1 = <fortran object>

VERSION
    1.23.0.dev0+1087.g0eaa40db3
```

So it seems to be fixed on master. Will close in a day or so if there are no updates.

**Comment by ketch** [comment]
In our experience the bug is not consistent. The function signature produced by f2py is different on different machines. But using numpy 1.18 always fixes it, and we've found no other way to fix it.

**Comment by HaoZeke** [comment]
Could you leave a reproducible example? Or a known machine where the incorrect results are produced? There should be no difference in the wrappers based on the underlying system.

**Comment by ketch** [comment]
> There should be no difference in the wrappers based on the underlying system.

I think we all agree on that!  Some characteristics of machines with this problem are mentioned here: https://github.com/clawpack/pyclaw/issues/681#issuecomment-1099947487

Let me know what else you need to know.

**Comment by WarrenWeckesser** [comment]
For what its worth, I experimented with this simplified version of the `step1` subroutine:
```
subroutine foo(num_eqn, num_ghost, mx, q)

    implicit double precision (a-h,o-z)
    double precision :: q(num_eqn, 1-num_ghost:mx+num_ghost)

!f2py intent(in,out) q
!f2py intent(in) num_eqn
!f2py intent(in) num_ghost
!f2py intent(in) mx

    return
end subroutine foo
```
I ran `f2py` with the command
```
    f2py --build-dir . -m foo foo.f90
```

With numpy 1.22.0 - 1.22.3, the signature of the wrapper generated by f2py is
```
    q = foo(q,[num_eqn,num_ghost,mx])
```
This is the unexpected result that is reported in this issue.

With the main development branch ('1.23.0.dev0+1087.g0eaa40db3'), and with all the older versions that I tested (1.19.0, 1.19.5, 1.20.0-1.20.3, 1.21.0-1.21.6), the signature is
```
    q = foo(num_ghost,mx,q,[num_eqn])
```
This is on a Linux machine, running Python 3.9.7 (miniconda), with all numpy versions except the main development branch installed with pip.

So for this platform and Python 3.9.7, it appears that only the 1.22 release series has the issue.
